upb/util/compare_test.cc

```diff
@@ -0,0 +1,255 @@
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Google LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL Google LLC BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "upb/util/compare.h"
+
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include <stdint.h>
+#include <vector>
+#include <string_view>
+
+struct UnknownField;
+
+using UnknownFields = std::vector<UnknownField>;
+
+enum class UnknownFieldType {
+  kVarint,
+  kLongVarint,  // Over-encoded to have distinct wire format.
+  kDelimited,
+  kFixed64,
+  kFixed32,
+  kGroup,
+};
+
+union UnknownFieldValue {
+    uint64_t varint;
+    uint64_t fixed64;
+    uint32_t fixed32;
+    // NULL-terminated (strings must not have embedded NULL).
+    const char* delimited;
+    UnknownFields* group;
+};
+
+struct TypeAndValue {
+  UnknownFieldType type;
+  UnknownFieldValue value;
+};
+
+struct UnknownField {
+  uint32_t field_number;
+  TypeAndValue value;
+};
+
+
+TypeAndValue Varint(uint64_t val) {
+  TypeAndValue ret{UnknownFieldType::kVarint};
+  ret.value.varint = val;
+  return ret;
+}
+
+TypeAndValue LongVarint(uint64_t val) {
+  TypeAndValue ret{UnknownFieldType::kLongVarint};
+  ret.value.varint = val;
+  return ret;
+}
+
+TypeAndValue Fixed64(uint64_t val) {
+  TypeAndValue ret{UnknownFieldType::kFixed64};
+  ret.value.fixed64 = val;
+  return ret;
+}
+
+TypeAndValue Fixed32(uint32_t val) {
+  TypeAndValue ret{UnknownFieldType::kFixed32};
+  ret.value.fixed32 = val;
+  return ret;
+}
+
+TypeAndValue Delimited(const char* val) {
+  TypeAndValue ret{UnknownFieldType::kDelimited};
+  ret.value.delimited = val;
+  return ret;
+}
+
+TypeAndValue Group(UnknownFields nested) {
+  TypeAndValue ret{UnknownFieldType::kGroup};
+  ret.value.group = &nested;
+  return ret;
+}
+
+void f() {
```

Comment:
what is the point of this function?

Reply:
Whoops, a test function left in. Deleted.